libexec/scoop-which.ps1

```diff
@@ -19,10 +19,14 @@ $path = "$($gcm.path)"
 $usershims = "$(resolve-path $(shimdir $false))"
 $globalshims = fullpath (shimdir $true) # don't resolve: may not exist
 
-if($path.endswith(".ps1") -and ($path -like "$usershims*" -or $path -like "$globalshims*")) {
-    $shimtext = Get-Content $path
-
-    $exepath = ($shimtext | Where-Object { $_.startswith('$path') }).split(' ') | Select-Object -Last 1 | Invoke-Expression
+if($path -like "$usershims*" -or $path -like "$globalshims*") {
+    $exepath = if ($path.endswith(".exe")) {
+        (Get-Content "$(join-path ([system.io.fileinfo]$path).DirectoryName ([system.io.fileinfo]$path).BaseName).shim" | Select-Object -First 1).replace('path = ', '')
+    } elseif ($path.endswith(".ps1")) {
+        (Get-Content $path | Where-Object { $_.startswith('$path') }).split(' ') | Select-Object -Last 1 | Invoke-Expression
+    } else {
+        # need to handle .cmd files
```

Comment:
```suggestion
        (Select-String -Path $path -Pattern '[''"]([^&=]*?)[''"]').Matches.Groups[1].Value
```

An ugly job but should work.

Reply:
Hmm, I'm not very strong at Regex but will this work for all types of batch files we create?

We currently have these types:

- https://github.com/ScoopInstaller/Scoop/blob/968e38bad8bd9cf7d1bf33bd0ba4c8b32d3db2f6/lib/core.ps1#L578
- https://github.com/ScoopInstaller/Scoop/blob/968e38bad8bd9cf7d1bf33bd0ba4c8b32d3db2f6/lib/core.ps1#L608
- https://github.com/ScoopInstaller/Scoop/blob/968e38bad8bd9cf7d1bf33bd0ba4c8b32d3db2f6/lib/core.ps1#L614
- https://github.com/ScoopInstaller/Scoop/blob/968e38bad8bd9cf7d1bf33bd0ba4c8b32d3db2f6/lib/core.ps1#L620
- https://github.com/ScoopInstaller/Scoop/blob/968e38bad8bd9cf7d1bf33bd0ba4c8b32d3db2f6/lib/core.ps1#L631

Reply:
I'm not sure if it works for `bash`, but the first four are okay with this regex.

If PATH in `.cmd` is the only one that enclosed in `"` or `'` and doesn't contain `&` and `=`, the regex will extract it.

Examples:

scoop.cmd (excluding `&` and `=` is for such `ps1` script)
```cmd
@echo off
setlocal enabledelayedexpansion
set args=%*
:: replace problem characters in arguments
set args=%args:"='%
set args=%args:(=`(%
set args=%args:)=`)%
set invalid="='
if !args! == !invalid! ( set args= )
powershell -noprofile -ex unrestricted "& 'D:\Scoop\apps\scoop\current\bin\scoop.ps1'  %args%;exit $lastexitcode"
```

idle.cmd
```cmd
@"D:\Scoop\apps\python39\current\Lib\idlelib\idle.bat"  %*
```

shinyproxy.cmd (`java`, `python` is the same, maybe `bash`?)
```cmd
@java -jar "D:\Scoop\apps\shinyproxy\current\shinyproxy.jar"  %*
```

Reply:
Oops, I forgot to put the bash path in double quotes. Fixed now.

Maybe we can add a negation in the regex so that it doesn't match the path with 'bash.exe' in it?

Reply:
Of course we could, but I think a better way is using `REM xxx` in the first line of every cmd so this function could find it without any additional job.

Since you are modifying shim creation function now, it is just the place to add them.

Reply:
> a better way is using REM xxx in the first line of every cmd so this function could find it without any additional job

I was thinking of the same thing yesterday!

But I was hesitant to do it because it required modifying all the shims. If this is fine, I will make this change.

Reply:
I think it's fine to add this comment, maybe in `.cmd` first, and then `.ps1`. `.shim` has a simple structure and needn't this.

Reply:
Done.

Reply:
```powershell
(Select-String -Path $path -Pattern '^(?:@rem|#)\s*(.*)$').Matches.Groups[1].Value
```

RegEx is `^(?:@rem|#)\s*(.*)$`, use `^` and `$` to find line begin and end. Please try it to see if works.

Reply:
Working!